src/vanilla/Model/CodeModelTS.cs

```diff
@@ -391,6 +394,110 @@ public string RequiredConstructorParametersTS
+        }
+
+        public virtual string GenerateContextConstructor(string emptyLine)
+        {
+            Action<TSBlock> guardCheck = block =>
+            {
+                foreach (var param in RequiredParameters)
+                {
+                    block.If($"{param.Name} === null || {param.Name} === undefined", then =>
+                    {
+                        then.Line($"throw new Error(\"\'{param.Name}\' cannot be null.\");");
+                    });
+                }
+
+                if (RequiredParameters.Any())
+                {
+                    block.Line(emptyLine);
+                }
+
+                block.If("!options", then =>
+                {
+                    then.Line("options = {};");
+                });
+
+                block.Line(emptyLine);
+                block.If("!options.userAgent", then =>
+                {
+                    then.ConstObjectVariable("defaultUserAgent", "msRest.getDefaultUserAgentValue()");
+                    then.Assignment("options.userAgent", "`${packageName}/${packageVersion} ${defaultUserAgent}`");
+                });
+
+                block.Line(GenerateCustomServiceClientOptions(emptyLine));
+                block.Line(emptyLine);
+            };
+
+            string credentialsParameter = RequiredParameters.Any(p => p.ModelType.IsPrimaryType(KnownPrimaryType.Credentials)) ? "credentials" : "undefined";
+            string superParameters = credentialsParameter + ", options";
+
+            Action<TSBlock> implementation = block =>
+            {
+                block.Line(emptyLine);
+                foreach (var property in Properties.Where(p => p.DefaultValue != null))
+                {
+                    block.ThisAssignment(property.Name, property.DefaultValue);
+                }
+
+                block.Line(GenerateBaseUri());
+
+                if (!String.IsNullOrEmpty(RequestContentType))
+                {
+                    block.ThisAssignment("requestContentType", $"\"{RequestContentType}\"");
+                }
+
+                foreach (var param in RequiredParameters)
+                {
+                    block.ThisAssignment(param.Name, param.Name);
+                }
+
+                if (OptionalParameters.Where(p => p.Name != "generatedClientRequestId").Any())
+                {
+                    foreach (var property in OptionalParameters.Where(p => p.Name != "generatedClientRequestId"))
+                    {
+                        block.If($"options.{property.Name} !== null && options.{property.Name} !== undefined", then =>
+                        {
+                            block.ThisAssignment(property.Name, $"options.{property.Name}");
+                        });
+                    }
+                }
+            };
+
+            return this.GenerateConstructor(superParameters, guardCheck, implementation);
```

Comment:
Rather than have two Actions in this argument list, I'd prefer to see `guardCheck` go into `implementation` and just have a single Action. Then you could inline the action and it would look more like you were building the code here.
Was there a reason you chose to go this route over a single action?

Reply:
so the `super(...)` call goes between those two, that was the level of abstraction I introduced to make it generic enough to work for both Context and Client